[Step-10] 장애 대응 프로세스

gyehwan24 opened the pull request:
### **커밋 설명**

| 커밋 | 설명 |
|------|------|
| `95ed24a` | 부하테스트 계획 문서 작성 (시나리오, 목표 TPS, 테스트 환경 정의) |
| `68a31dc` | 부하 테스트 결과 분석 및 배포 스펙 문서 작성 |
| `4aa6046` | 장애 대응 매뉴얼 및 발표 자료 작성 |
| `d918a74` | 토큰 발급 API 성능 개선 (에러율 3.3% → 0.01%) |
| `9f5acd8` | 성능 개선 결과 반영하여 문서 업데이트 |

---
### **리뷰 받고 싶은 내용(질문)**

- **커밋**: `d918a74` (`QueueRedisRepository.java`)
- **내용**: Redis Lua 스크립트로 대기열 추가 로직을 원자화했습니다. 기존에는 `ZSCORE` 조회 → `ZADD` 추가가 분리되어 Race Condition이 발생했는데, Lua 스크립트로 단일 트랜잭션화한 접근이 적절한지 검토 부탁드립니다.

- **커밋**: `d918a74` (`TokenExceptionHandler.java`)
- **내용**: 중복 대기열 진입 시 `IllegalStateException`을 HTTP 409(Conflict)로 변환하도록 예외 핸들러를 추가했습니다. 비즈니스 예외와 시스템 에러를 구분하는 이 설계가 적절한지 리뷰 부탁드립니다.

---

### **과제 셀프 피드백**

**모호했던 부분**
- 동시 사용자 수 산정 기준: 실제 트래픽 데이터 없이 적정 수치를 결정하기 어려움
- Thread Group 간 비율 설정
- 로컬 Docker 환경 테스트 결과를 운영 환경에 어디까지 적용할 수 있는지

**좋았던 부분**
- 실제 부하 테스트를 통해 병목 구간(토큰 발급 API)을 발견하고 개선까지 진행한 점
- 에러율 3.3% → 0.01%로 개선하며 실질적인 성능 향상을 달성

---

### 기술적 성장

**새로 학습한 개념**
- Redis Lua 스크립트를 활용한 원자적 연산 구현
- JMeter를 활용한 복합 시나리오 부하 테스트 (7개 Thread Group 연계)

**기존 지식의 심화**
- HikariCP, Redisson 커넥션 풀 튜닝과 고부하 시 동작 이해
- Race Condition의 실제 발생 사례와 해결 방법 체득

**기술적 도전과 해결**
- 고부하(730명) 테스트에서 토큰 API 에러율 3.3% 발생
- 원인: 대기열 추가 로직의 check-then-act 패턴에서 Race Condition 발생
- 해결: Lua 스크립트로 원자화 + 예외 핸들러로 비즈니스 예외 분류
- 결과: 에러율 0.01% 달성

gyehwan24 committed:
docs: 부하테스트 계획 문서 작성

gyehwan24 committed:
docs: 부하 테스트 결과 및 배포 스펙 문서 작성

- 부하 테스트 결과 보고서 작성 (load-test-result.md)
  - 기본 부하 (350 users): TPS 4,523, 에러율 0%
  - 고부하 (730 users): TPS 4,479, 에러율 3.3%
  - 토큰 발급 API 병목 식별
- 배포 스펙 권장안 작성 (deployment-spec.md)
  - 권장 동시 사용자: 500명
  - HikariCP 3 → 20 증가 권장
  - 환경별 리소스 스펙 정의

gyehwan24 committed:
docs: 장애 대응 매뉴얼 및 발표 자료 작성

- 장애 대응 매뉴얼 작성 (incident-response-manual.md)
  - 장애 등급 정의 (P1~P4)
  - 애플리케이션/DB/Redis 장애 대응 절차
  - 모니터링 체크리스트 및 알림 설정
  - 에스컬레이션 경로 정의
- 최종 발표 자료 작성 (final-presentation.md)
  - 시스템 아키텍처 개요
  - 성능 테스트 결과 요약
  - 시스템 수용량 및 스케일링 전략
  - 취약점 분석 및 개선 방향
- JMeter 테스트 스크립트 추가 (jmeter/)
  - concert-full-load-test.jmx: 전체 API 부하 테스트
  - run-load-test.sh: 실행 스크립트

gyehwan24 committed:
perf: 토큰 발급 API 성능 개선 (에러율 3.3% → 0.01%)

- HikariCP 커넥션 풀 증가: 3 → 20
- Redisson 커넥션 풀 증가: 50 → 100
- 스케줄러 스레드 풀 증가: 2 → 5
- addToWaitingQueue Lua 스크립트 원자화로 Race Condition 해결
- TokenExceptionHandler 추가로 중복 요청 시 HTTP 409 응답
- JMeter Assertion 수정: 200, 409 모두 성공으로 처리

gyehwan24 committed:
docs: 성능 개선 결과 반영하여 문서 업데이트

- load-test-result.md: 개선 후 재테스트 결과 섹션 추가 (에러율 3.3% → 0.01%)
- deployment-spec.md: 적용 완료된 설정 및 검증된 성능 반영
- final-presentation.md: 취약점 해결 완료 상태로 업데이트

khj68 commented:
안녕하세요 계환님. 이번 과제 진행하느라 고생 많으셨습니다.
부하 테스트를 통해 실제 병목 지점을 발견하고, Lua 스크립트로 Race Condition을 해결하여 에러율을 3.3%에서 0.01%로 개선하신 점이 인상적입니다. 특히 문제를 발견하고 해결하는 전 과정을 문서화하신 것도 좋습니다.

질문주신 2가지 내용을 중심으로 구체적인 피드백 남겨드리니 참고해주세요.

---

## 리뷰 포인트에 대한 답변

### 1. Redis Lua 스크립트로 대기열 추가 로직 원자화

**Q. Redis Lua 스크립트로 대기열 추가 로직을 원자화했습니다. 기존에는 `ZSCORE` 조회 → `ZADD` 추가가 분리되어 Race Condition이 발생했는데, Lua 스크립트로 단일 트랜잭션화한 접근이 적절한지 검토 부탁드립니다.**

완벽한 접근입니다. Lua 스크립트를 사용한 원자화는 Redis에서 Race Condition을 해결하는 표준 방법이에요.

계환님이 구현하신 Lua 스크립트를 보면요.

```lua
local waitingKey = KEYS[1]
local activeKey = KEYS[2]
local userId = ARGV[1]
local score = tonumber(ARGV[2])

-- 대기열에 이미 있는지 확인
local inWaiting = redis.call('ZSCORE', waitingKey, userId)
if inWaiting then
    return -1  -- 이미 대기열에 있음
end

-- 활성 큐에 있는지 확인 (만료 여부도 체크)
local expireAt = redis.call('ZSCORE', activeKey, userId)
if expireAt and tonumber(expireAt) > score then
    return -2  -- 이미 활성 상태
end

-- 대기열에 추가
redis.call('ZADD', waitingKey, score, userId)

-- 순번 반환 (0-based)
return redis.call('ZRANK', waitingKey, userId)
```

이 스크립트는 세 가지 작업을 원자적으로 처리합니다.

1. 중복 체크 (대기열)
2. 활성 상태 체크 (활성 큐)
3. 추가 및 순번 반환

**왜 이 방식이 적절한가요?**

Redis는 단일 스레드로 명령을 처리합니다. Lua 스크립트는 실행 중에 다른 명령이 끼어들 수 없어서, 여러 클라이언트가 동시에 요청해도 Race Condition이 발생하지 않아요.

**기존 코드의 문제점**

```java
// 기존 코드 - Race Condition 발생 가능
if (isInWaitingQueue(userId)) {
    throw new IllegalStateException("이미 대기열에 있습니다.");
}
if (isActiveUser(userId)) {
    throw new IllegalStateException("이미 활성화된 상태입니다.");
}
redisTemplate.opsForZSet().add(WAITING_QUEUE_KEY, userIdStr, score);
```

이 코드는 세 번의 네트워크 왕복이 발생하고, 각 단계 사이에 다른 요청이 끼어들 수 있어요.

**시나리오: Race Condition 발생**

```
시간 | 스레드 A                    | 스레드 B
-----|---------------------------|---------------------------
T1   | ZSCORE 확인 (없음)         |
T2   |                            | ZSCORE 확인 (없음)
T3   |                            | ZADD 추가
T4   | ZADD 추가 (중복!)          |
```

이런 상황에서 두 스레드 모두 대기열에 추가되어 중복이 발생합니다.

**Lua 스크립트의 장점**

1. 원자성 보장: 스크립트 전체가 하나의 명령으로 실행
2. 네트워크 왕복 감소: 3번 → 1번
3. 성능 향상: 네트워크 지연 감소로 응답 시간 단축

**실무에서도 이 방식을 많이 사용해요**

Redis에서 원자적 연산이 필요한 경우 Lua 스크립트를 사용하는 것이 일반적입니다. 예를 들어 재고 차감, 랭킹 업데이트, 분산 카운터 등에서 활용됩니다.

**추가 개선 제안**

현재 구현도 충분히 좋지만, 스크립트 로드 최적화를 고려해볼 수 있어요.

```java
@Repository
public class QueueRedisRepository {
    private final StringRedisTemplate redisTemplate;
    private final DefaultRedisScript<Long> addToWaitingQueueScript;
    private String scriptSha;  // SHA1 해시 캐싱
    
    @PostConstruct
    public void loadScript() {
        // 스크립트를 미리 로드하여 SHA1 해시 저장
        this.scriptSha = redisTemplate.execute(
            (RedisCallback<String>) connection -> {
                byte[] scriptBytes = addToWaitingQueueScript.getScriptAsString().getBytes();
                return connection.scriptLoad(scriptBytes);
            }
        );
    }
    
    public long addToWaitingQueue(Long userId) {
        try {
            // EVALSHA로 SHA1만 전송 (스크립트 전체 대신 40바이트만)
            Long result = redisTemplate.execute(
                new DefaultRedisScript<>(scriptSha, Long.class),
                keys, args
            );
            // ...
        } catch (RedisScriptException e) {
            if (e.getMessage().contains("NOSCRIPT")) {
                // 스크립트가 없으면 다시 로드 후 재시도
                loadScript();
                return addToWaitingQueue(userId);
            }
            throw e;
        }
    }
}
```

이렇게 하면 스크립트 전체(수백 바이트) 대신 SHA1 해시(40바이트)만 전송해서 네트워크 트래픽을 줄일 수 있어요. 다만 현재 구현도 충분히 좋으니 필수는 아닙니다.

---

### 2. 비즈니스 예외와 시스템 에러 구분

**Q. 중복 대기열 진입 시 `IllegalStateException`을 HTTP 409(Conflict)로 변환하도록 예외 핸들러를 추가했습니다. 비즈니스 예외와 시스템 에러를 구분하는 이 설계가 적절한지 리뷰 부탁드립니다.**

매우 적절한 설계입니다. HTTP 상태 코드를 비즈니스 의미에 맞게 매핑하는 것은 RESTful API 설계의 핵심이에요.

계환님이 구현하신 예외 핸들러를 보면요.

```java
@RestControllerAdvice(assignableTypes = TokenController.class)
public class TokenExceptionHandler {

    @ExceptionHandler(IllegalStateException.class)
    public ResponseEntity<ErrorResponse> handleIllegalState(IllegalStateException e) {
        log.debug("Token API IllegalStateException: {}", e.getMessage());
        return ResponseEntity.status(HttpStatus.CONFLICT)
                .body(new ErrorResponse("DUPLICATE_REQUEST", e.getMessage()));
    }
}
```

**왜 HTTP 409 Conflict가 적절한가요?**

RFC 7231에 따르면, 409 Conflict는 "요청이 현재 리소스의 상태와 충돌할 때" 사용합니다. 중복 대기열 진입은 사용자가 이미 대기열에 있는 상태에서 다시 진입을 시도하는 것이므로, 현재 상태와 충돌하는 상황이에요.

**예외 분류의 중요성**

비즈니스 예외와 시스템 에러를 구분하면 클라이언트가 적절히 대응할 수 있어요.

| 예외 유형 | HTTP 상태 코드 | 클라이언트 대응 |
|----------|--------------|----------------|
| 중복 대기열 진입 | 409 Conflict | 재시도 불필요, 상태 확인 |
| 잘못된 파라미터 | 400 Bad Request | 요청 수정 후 재시도 |
| 서버 오류 | 500 Internal Server Error | 재시도 또는 에러 리포트 |

**현재 설계의 장점**

1. **도메인별 예외 핸들러**: `assignableTypes = TokenController.class`로 Token API에만 적용되어, 다른 도메인의 `IllegalStateException`과 충돌하지 않아요.

2. **명확한 에러 코드**: `DUPLICATE_REQUEST` 같은 코드를 제공해서 클라이언트가 에러 타입을 구분할 수 있어요.

3. **적절한 로깅 레벨**: `log.debug`를 사용해서 정상적인 비즈니스 플로우는 로그를 남기지 않아요.

**추가 개선 제안**

현재 설계도 충분히 좋지만, 더 세밀한 제어를 원한다면 전용 예외 클래스를 만들 수도 있어요.

```java
// 전용 예외 클래스
public class DuplicateQueueEntryException extends RuntimeException {
    private final Long userId;
    
    public DuplicateQueueEntryException(Long userId) {
        super("이미 대기열에 있습니다.");
        this.userId = userId;
    }
}

// 예외 핸들러
@ExceptionHandler(DuplicateQueueEntryException.class)
public ResponseEntity<ErrorResponse> handleDuplicateQueueEntry(
        DuplicateQueueEntryException e) {
    return ResponseEntity.status(HttpStatus.CONFLICT)
            .body(new ErrorResponse("DUPLICATE_REQUEST", e.getMessage()));
}

// 서비스 코드
if (result == -1) {
    throw new DuplicateQueueEntryException(userId);
}
```

이렇게 하면 `IllegalStateException`의 다른 용도(예: 활성화된 상태)와 구분할 수 있어요. 다만 현재처럼 메시지로 구분하는 것도 충분히 실용적입니다.

**주의할 점**

현재 `@RestControllerAdvice(assignableTypes = TokenController.class)`로 범위를 제한하신 점이 좋습니다. 만약 전역 핸들러가 있다면, 더 구체적인 핸들러가 우선 적용되므로 문제없어요.

---

## 잘한 점

### 1. 부하 테스트를 통한 실제 문제 발견

고부하(730명) 테스트에서 에러율 3.3%를 발견하고, 원인을 정확히 파악하신 점이 좋습니다. 단순히 코드를 개선하는 것이 아니라, 실제 성능 문제를 측정하고 해결하는 접근이 실무와 동일해요.

### 2. Lua 스크립트로 Race Condition 해결

`ZSCORE` 조회와 `ZADD` 추가를 분리했을 때 발생하는 Race Condition을 Lua 스크립트로 해결하신 점이 정확합니다. Redis의 단일 스레드 특성을 활용한 원자적 연산 구현이 표준적인 해결 방법이에요.

### 3. 예외 핸들러로 HTTP 상태 코드 적절히 매핑

비즈니스 예외(`IllegalStateException`)를 HTTP 409로 변환하신 점이 좋습니다. 클라이언트가 에러 타입을 구분할 수 있어서 적절한 대응이 가능해요.

### 4. 문서화가 체계적입니다

부하 테스트 계획, 결과 분석, 배포 스펙, 장애 대응 매뉴얼까지 작성하신 점이 인상적입니다. 특히 실제 테스트 결과를 바탕으로 배포 스펙을 제안하신 부분이 실무적이에요.

### 5. 성능 개선 결과가 명확합니다

에러율 3.3% → 0.01%로 개선한 결과가 구체적이고, 개선 전후 비교가 명확합니다. 이런 정량적 결과는 리뷰어가 개선 효과를 쉽게 파악할 수 있게 해요.

---

## 추가 개선 제안

### 1. Lua 스크립트 에러 처리 강화

현재 스크립트에서 `result == null`일 때만 예외를 던지고 있는데, 스크립트 실행 자체가 실패할 수 있는 경우도 고려하면 좋을 것 같아요.

```java
public long addToWaitingQueue(Long userId) {
    try {
        Long result = redisTemplate.execute(
                addToWaitingQueueScript,
                keys,
                userIdStr,
                String.valueOf(score)
        );

        if (result == null) {
            throw new IllegalStateException("대기열 추가 실패");
        }

        if (result == -1) {
            throw new IllegalStateException("이미 대기열에 있습니다.");
        }

        if (result == -2) {
            throw new IllegalStateException("이미 활성화된 상태입니다.");
        }

        return result + 1;
    } catch (RedisException e) {
        log.error("Redis 스크립트 실행 실패: userId={}", userId, e);
        throw new IllegalStateException("대기열 처리 중 오류가 발생했습니다.", e);
    }
}
```

### 2. 예외 메시지 일관성

현재 `IllegalStateException`의 메시지가 "이미 대기열에 있습니다"와 "이미 활성화된 상태입니다"로 구분되어 있는데, 클라이언트가 더 명확하게 구분할 수 있도록 에러 코드를 활용하는 것도 좋아요.

```java
@ExceptionHandler(IllegalStateException.class)
public ResponseEntity<ErrorResponse> handleIllegalState(IllegalStateException e) {
    String errorCode = determineErrorCode(e.getMessage());
    return ResponseEntity.status(HttpStatus.CONFLICT)
            .body(new ErrorResponse(errorCode, e.getMessage()));
}

private String determineErrorCode(String message) {
    if (message.contains("대기열")) {
        return "DUPLICATE_QUEUE_ENTRY";
    } else if (message.contains("활성화")) {
        return "ALREADY_ACTIVE";
    }
    return "DUPLICATE_REQUEST";
}
```

다만 현재 구현도 충분히 실용적이니 필수는 아닙니다.

### 3. 스크립트 검증 로직 추가

Lua 스크립트에서 반환값이 예상 범위를 벗어나는 경우를 검증하면 더 안전해요.

```lua
local waitingKey = KEYS[1]
local activeKey = KEYS[2]
local userId = ARGV[1]
local score = tonumber(ARGV[2])

-- 입력 검증
if not userId or not score then
    return -3  -- INVALID_INPUT
end

-- 대기열에 이미 있는지 확인
local inWaiting = redis.call('ZSCORE', waitingKey, userId)
if inWaiting then
    return -1  -- 이미 대기열에 있음
end

-- 활성 큐에 있는지 확인 (만료 여부도 체크)
local expireAt = redis.call('ZSCORE', activeKey, userId)
if expireAt and tonumber(expireAt) > score then
    return -2  -- 이미 활성 상태
end

-- 대기열에 추가
redis.call('ZADD', waitingKey, score, userId)

-- 순번 반환 (0-based)
local rank = redis.call('ZRANK', waitingKey, userId)
if rank == nil then
    return -4  -- 추가 실패
end

return rank
```

---

## 전체 평가

부하 테스트를 통해 실제 문제를 발견하고, Lua 스크립트로 Race Condition을 해결하여 에러율을 대폭 개선하신 점이 인상적입니다. 특히 문제 해결 과정을 문서화하고, 배포 스펙과 장애 대응 매뉴얼까지 작성하신 점이 실무적이에요.

**잘한 부분**
- 부하 테스트를 통한 실제 문제 발견 및 해결
- Lua 스크립트로 Race Condition 해결
- 비즈니스 예외를 적절한 HTTP 상태 코드로 매핑
- 체계적인 문서화 (테스트 계획, 결과, 배포 스펙, 장애 대응 매뉴얼)
- 정량적 성능 개선 결과 (에러율 3.3% → 0.01%)

**개선이 필요한 부분**
- Lua 스크립트 에러 처리 강화 (RedisException 처리)
- 예외 메시지 기반 에러 코드 분류 개선 (선택사항)

과제 진행하느라 고생 많으셨습니다!

gyehwan24 committed:
feat: Redis 대기열 안정성 개선

- Lua 스크립트 입력값 검증 추가 (userId, score nil 체크)
- ZADD 후 ZRANK 결과 nil 체크로 추가 실패 감지
- Redis 예외 처리 추가 (연결 실패, 명령 실행 실패)
- 에러 코드 세분화로 명확한 오류 원인 파악 가능

gyehwan24 commented:
### 피드백 반영 개선사항 적용
1. Lua 스크립트 입력값 검증

```lua
  -- userId 검증
  if not userId or userId == '' then
      return -3  -- INVALID_USER_ID
  end

  -- score 검증
  if not score then
      return -4  -- INVALID_SCORE
  end

  -- ZADD 후 결과 검증
  if rank == nil then
      return -5  -- ADD_FAILED
  end
```

  2. Redis 예외 처리

```java
  catch (RedisConnectionFailureException e) {
      log.error("Redis 연결 실패: userId={}", userId, e);
      throw new IllegalStateException("대기열 서비스를 일시적으로 사용할 수 없습니다.", e);
  }
  catch (DataAccessException e) {
      log.error("Redis 명령 실행 실패: userId={}", userId, e);
      throw new IllegalStateException("대기열 처리 중 오류가 발생했습니다.", e);
  }
```

  3. 에러 코드 매핑

  | 코드  | 의미             | 예외                       |
  |-----|----------------|--------------------------|
  | -1  | 이미 대기열에 있음     | IllegalStateException    |
  | -2  | 이미 활성 상태       | IllegalStateException    |
  | -3  | 유효하지 않은 userId | IllegalArgumentException |
  | -4  | 유효하지 않은 score  | IllegalArgumentException |
  | -5  | 대기열 추가 실패      | IllegalStateException    |
